src/common/intrusive_lru.h

```diff
@@ -0,0 +1,174 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <boost/intrusive_ptr.hpp>
+#include <boost/intrusive/set.hpp>
+#include <boost/intrusive/list.hpp>
+#include <boost/smart_ptr/intrusive_ref_counter.hpp>
+
+#include "include/ceph_assert.h"
+
+namespace ceph::common {
+
+template <typename K, typename V, typename VToK>
+struct intrusive_lru_config {
+  using key_type = K;
+  using value_type = V;
+  using key_of_value = VToK;
+};
+
+template <typename Config>
+class intrusive_lru;
+
+template <typename Config>
+class intrusive_lru_base;
+
+template <typename Config>
+void intrusive_ptr_add_ref(intrusive_lru_base<Config> *p);
+
+template <typename Config>
+void intrusive_ptr_release(intrusive_lru_base<Config> *p);
+
+
+template <typename Config>
+class intrusive_lru_base {
+  unsigned use_count = 0;
+
+  // null if unreferenced
+  intrusive_lru<Config> *lru = nullptr;
+
+public:
+  boost::intrusive::set_member_hook<> set_hook;
+  boost::intrusive::list_member_hook<> list_hook;
+
+  using Ref = boost::intrusive_ptr<typename Config::value_type>;
```

Comment:
nit, might want to s/Ref/value_ref_t/ to be more specific, as `Ref` without more context could be confusing.

Reply:
It's scoped to the type, so it would always be qualified outside of this file.

Reply:
i found myself lost when reading `ObjectContext`. where we have `Ref head`. so it took me a while searching for `Ref` for its definition, that's why i thought it might be a good idea to qualify it a little bit.

Reply:
I'd really rather leave it like this as it follows a convention we use pretty often:

ceph/src [sjust/wip-crimson-obc-errorator-perf●] » git grep 'using Ref =' | cat
common/intrusive_lru.h:  using Ref = boost::intrusive_ptr<typename Config::value_type>;
crimson/common/type_helpers.h:template<typename T> using Ref = boost::intrusive_ptr<T>;
crimson/mgr/client.h:template<typename Message> using Ref = boost::intrusive_ptr<Message>;
crimson/mon/MonClient.h:template<typename Message> using Ref = boost::intrusive_ptr<Message>;
crimson/osd/heartbeat.h:template<typename Message> using Ref = boost::intrusive_ptr<Message>;
crimson/osd/osd_meta.h:  template<typename T> using Ref = boost::intrusive_ptr<T>;
os/memstore/MemStore.h:    using Ref = ceph::ref_t<Object>;
osd/scheduler/OpSchedulerItem.h:    using Ref = unique_ptr<OrderLocker>;
osd/scheduler/OpSchedulerItem.h:    using Ref = std::unique_ptr<OpQueueable>;
tools/crimson/perf_crimson_msgr.cc:using Ref = boost::intrusive_ptr<Message>;